File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        var islands = 0;
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[0].length; j++){
                if(grid[i][j] == '1'){
                    dfs_visit(grid, i, j);
                    islands++;
                }
            }
        }
        return islands;
    }
    private void dfs_visit(char[][] grid, int r, int c){
        if(grid[r][c] == '1'){
            grid[r][c] = '0';
            if(r+1 < grid.length) dfs_visit(grid, r+1, c);
            if(c+1 < grid[0].length) dfs_visit(grid, r, c+1);
            if(r >= 1) dfs_visit(grid, r-1, c);
            if(c >= 1) dfs_visit(grid, r, c-1);
        }
    }
}

File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int islands = 0;
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(grid[i][j] == '1'){
                    dfs(grid, i, j);
                    islands++;
                }
            }
        }
        return islands;
    }
    
    void dfs(vector<vector<char>>& g, int i, int j){
        if(g[i][j] == '1'){
            g[i][j] = '0';
            if(i >= 1){
                dfs(g, i-1, j);
            }
            if(j >= 1){
                dfs(g, i, j-1);
            }
            if(i < g.size()-1){
                dfs(g, i+1, j);
            }
            if(j < g[0].size()-1){
                dfs(g, i, j+1);
            }
        }
    }
};